Source/FieldSolver/MagnetostaticSolver/MagnetostaticSolver.cpp

```diff
@@ -0,0 +1,611 @@
+#include <AMReX_GpuLaunch.H>
+#include <AMReX_GpuQualifiers.H>
+#include <AMReX_IndexType.H>
+#include <AMReX_IntVect.H>
+#include <AMReX_LO_BCTYPES.H>
+#include <AMReX_MFIter.H>
+#include <AMReX_MLMG.H>
+#include <AMReX_MultiFab.H>
+#include <AMReX_REAL.H>
+#include <AMReX_SPACE.H>
+#include <AMReX_Vector.H>
+#include <AMReX_MFInterp_C.H>
+#ifdef AMREX_USE_EB
+#   include <AMReX_EBFabFactory.H>
+#endif
+
+#include <array>
+#include <memory>
+#include <string>
+
+using namespace amrex;
+
+void
+WarpX::ComputeCurrentDensityField()
+{
+    WARPX_PROFILE("WarpX::ComputeCurrentDensityField");
+    // Fields have been reset in Electrostatic solver for this time step, these fields
+    // are added into the E & B fields after electrostatic solve
+
+    WARPX_ALWAYS_ASSERT_WITH_MESSAGE(do_current_centering == true,
+                                     "Error: Magnetostatic solver requires warpx.do_current_centering=true");
+
+    AddCurrentDensityFieldLabFrame();
+
+    // Transfer fields from 'fp' array to 'aux' array.
+    // This is needed when using momentum conservation
+    // since they are different arrays in that case.
+    UpdateAuxilaryData();
+    FillBoundaryAux(guard_cells.ng_UpdateAux);
+
+}
+
+void
+WarpX::AddCurrentDensityFieldLabFrame()
+{
+    WARPX_PROFILE("WarpX::AddCurrentDensityFieldLabFrame");
+
+    // Store the boundary conditions for the field solver if they haven't been
+    // stored yet
+    if (!m_vector_poisson_boundary_handler.bcs_set) {
+        m_vector_poisson_boundary_handler.defineVectorPotentialBCs();
+    }
+
+#ifdef WARPX_DIM_RZ
+    WARPX_ALWAYS_ASSERT_WITH_MESSAGE(n_rz_azimuthal_modes == 1,
+                                     "Error: RZ magnetostatic only implemented for a single mode");
+#endif
+
+    // reset current_fp before depositing current density for this step
+    for (int lev = 0; lev <= max_level; lev++) {
+        for (int dim=0; dim < 3; dim++) {
+            current_fp_nodal[lev][dim]->setVal(0.);
+        }
+    }
+
+    // Deposit current density (source of Poisson solver)
+    for (int ispecies=0; ispecies<mypc->nSpecies(); ispecies++){
+        WarpXParticleContainer& species = mypc->GetParticleContainer(ispecies);
+        if (!species.do_not_deposit) {
+            species.DepositCurrent(current_fp_nodal, dt[0], 0.);
+        }
+    }
+
+    // This will interpolate to staggered grid, filter, and synchronize
+    SyncCurrent(current_fp, current_cp); // Apply filter, perform MPI exchange, interpolate across levels
+
+#ifdef WARPX_DIM_RZ
+    for (int lev = 0; lev <= max_level; lev++) {
+        ApplyInverseVolumeScalingToCurrentDensity(current_fp[lev][0].get(),
+                                                  current_fp[lev][1].get(),
+                                                  current_fp[lev][2].get(), lev);
+    }
+#endif
+
+    // NOTE:  (SEC 12/2/22)
+    // Interpolate fields back to nodal current
+    // This could be done faster, but want to check if answer is right
```

Comment:
Yes, I think that the code could be potentially simplified by adding the following block here:
https://github.com/ECP-WarpX/WarpX/blob/development/Source/WarpX.cpp#L1838
```
if (WarpX::electrostatic_solver == RelativisticMagnetostatic) {
        jx_nodal_flag  = IntVect::TheNodeVector();
        jy_nodal_flag  = IntVect::TheNodeVector();
        jz_nodal_flag  = IntVect::TheNodeVector();
}
```
In this case, all the operations in `AddCurrentDensityFieldLabFrame` could be done directly on `current_fp` (without ever having to touch `current_nodal_fp` and/or interpolate back and forth between `currenf_fp` and `current_nodal_fp`). We would also remove the need to set `do_current_centering`.

Reply:
So this scheme works well in 3D and eliminates interpolating back and forth between nodal and e-field staggered mesh locations..  The ApplyInverseVolumeScalingToCurrentDensity routine throws an Assertion error since it only works for cell edges.  It will need to be extended to work for Nodal currents for it to work correctly.

Reply:
So the weighting of Jr has been adjusted to allow for the nodal inverse current density scaling.  Please take a look.  Otherwise I think this should be implemented now.  Please take a look at the latest commit.